docs/zh-CN/src/config.md
# 配置文件

前面我们已经学会了，如果以最基本的方式启动chubaodb。下面我们来学一种非常高级的用法，通过配置文件和启动参数来启动chubaodb。在学这个之前我们可能需要了简单解下chubaodb的架构，chubaodb份了三个角色`master`,`pserver`,`router`.

* master 是集群管理，元数据管理的模块。表结构，数据均衡，failover 等
* pserver 是数据存储及计算节点。
* router 是提供了restful的数据查询组件。router 是无状态的。

下面是一个简单的config例子。也是默认config的参数

````
[global]
# the name will validate join cluster by same name
name = "chubaodb"
# your server ip for connect, you can use -i in rags to set it
# ip = "127.0.0.1"
# log path , If you are in a production environment, You'd better set absolute paths
log = "logs/"
# default log type for any model
log_level = "info"
# log file size for rolling
log_limit_bytes = 128000000
# number of reserved log files
log_file_count = 10
# Whether to use distributed storage. If it's true, there's only one
shared_disk = false

[router]
# port for server
http_port = 8080

[ps]
#set zone num, default is 0
zone = "default"
# you data save to disk path ,If you are in a production environment, You'd better set absolute paths
data = "data/"
# port for server
rpc_port = 9090
# how often to refresh the index
flush_sleep_sec = 3
    [ps.raft]
        heartbeat_port = 10030
        replicate_port = 10031
        # how size of num for memory
        log_max_num = 200000
        # how size of num for memory
        log_min_num = 100000
        # how size of num for memory
        log_file_size_mb = 128
        # Three  without a heartbeat , follower to begin consecutive elections
        heartbeate_ms = 500
    

[[masters]]
# master ip for service
ip = "127.0.0.1"
# port for server
http_port = 7070
# master data path for meta
data = "data/meta/"
````

可以发现我们把 三个模块写到了同一个配置文件。同时各个节点通过参数来选择启动的模块。实现了一个配置文件走天下的易用功能。

可以通过 `./chubaodb --help` 获得参数说明

````
(base) ➜  release git:(async-std) ✗ ./chubaodb --help
chubaodb 0.1.0
hello index world

USAGE:
    chubaodb [SUBCOMMAND]

FLAGS:
    -h, --help       Prints help information
    -V, --version    Prints version information

SUBCOMMANDS:
    all       
    help      Prints this message or the help of the given subcommand(s)
    master    
    ps        
    router   
````

启动可以通过 `./chubaodb all --help ` , `./chubaodb ps --help ` .....来获取更多。参数说明

* 比如我们想把三个模块在一个进程中启动 那么就是. `./chubaodb all -c ../../config/config.toml` .

````
./chubaodb all -c ../../config/config.toml
load config by path: ../../config/config.toml
2020-07-15 11:05:39 - INFO - chubaodb::util::config(189) - log init ok 
2020-07-15 11:05:39 - INFO - chubaodb(149) - All ChubaoDB servers were started successfully!
2020-07-15 11:05:39 - INFO - chubaodb::router::server(29) - router is listening on http://0.0.0.0:8080
2020-07-15 11:05:39 - INFO - actix_server::builder(262) - Starting 8 workers
2020-07-15 11:05:39 - INFO - chubaodb::util::http_client(21) - send get for url:http://127.0.0.1:7070/my_ip
2020-07-15 11:05:39 - INFO - surf::middleware::logger::native(119) - sending request
2020-07-15 11:05:39 - INFO - actix_server::builder(276) - Starting "actix-web-service-0.0.0.0:8080" service on 0.0.0.0:8080
2020-07-15 11:05:39 - WARN - isahc::handler(209) - request completed with error [id=AtomicCell { value: 0 }]: ConnectFailed: failed to connect to the server
2020-07-15 11:05:39 - ERROR - chubaodb::pserver::server(41) - got ip from master has err:Error(InternalErr, "ConnectFailed: failed to connect to the server")
2020-07-15 11:05:39 - INFO - chubaodb::master::server(43) - master listening on http://0.0.0.0:7070
2020-07-15 11:05:39 - INFO - actix_server::builder(262) - Starting 8 workers
2020-07-15 11:05:39 - INFO - actix_server::builder(276) - Starting "actix-web-service-0.0.0.0:7070" service on 0.0.0.0:7070
2020-07-15 11:05:40 - INFO - chubaodb::util::http_client(21) - send get for url:http://127.0.0.1:7070/my_ip
2020-07-15 11:05:40 - INFO - surf::middleware::logger::native(119) - sending request
2020-07-15 11:05:40 - INFO - chubaodb::master::server(440) - success_response [Object({"ip": String("127.0.0.1")})]
2020-07-15 11:05:40 - INFO - surf::middleware::logger::native(119) - request completed
2020-07-15 11:05:40 - INFO - chubaodb::pserver::server(36) - got my ip:127.0.0.1 from master
2020-07-15 11:05:40 - INFO - chubaodb::util::http_client(51) - send post for url:http://127.0.0.1:7070/pserver/register
2020-07-15 11:05:40 - INFO - surf::middleware::logger::native(119) - sending request
2020-07-15 11:05:40 - INFO - chubaodb::master::server(301) - prepare to heartbeat with address 127.0.0.1:9090, zone default
2020-07-15 11:05:40 - INFO - chubaodb::master::server(440) - success_response [PServer { id: Some(1), addr: "127.0.0.1:9090", write_partitions: [], zone: "default", modify_time: 0 }]
2020-07-15 11:05:40 - INFO - surf::middleware::logger::native(119) - request completed
2020-07-15 11:05:40 - INFO - chubaodb::pserver::service(111) - register to master ok: node_id:Some(1) 
2020-07-15 11:05:40 - INFO - chubaodb::pserver::service(126) - register server line:PServer { id: Some(1), addr: "127.0.0.1:9090", write_partitions: [], zone: "default", modify_time: 0 }
2020-07-15 11:05:40 - INFO - chubaodb::pserver::server(59) - init pserver OK use time:Ok(5.333ms)
````

* 比如我们只启动 router 那么就是. `./chubaodb router -c ../../config/config.toml` .

````
(base) ➜  release git:(async-std) ✗ ./chubaodb router -c ../../config/config.toml
load config by path: ../../config/config.toml
2020-07-15 11:01:59 - INFO - chubaodb::util::config(189) - log init ok 
2020-07-15 11:01:59 - INFO - chubaodb(149) - All ChubaoDB servers were started successfully!
2020-07-15 11:01:59 - INFO - chubaodb::router::server(29) - router is listening on http://0.0.0.0:8080
2020-07-15 11:01:59 - INFO - actix_server::builder(262) - Starting 8 workers
2020-07-15 11:01:59 - INFO - actix_server::builder(276) - Starting "actix-web-service-0.0.0.0:8080" service on 0.0.0.0:8080
````

* 比如我们只启动 master 那么就是. `./chubaodb master -c ../../config/config.toml` .

````
load config by path: ../../config/config.toml
2020-07-15 11:03:11 - INFO - chubaodb::util::config(189) - log init ok 
2020-07-15 11:03:11 - INFO - chubaodb(149) - All ChubaoDB servers were started successfully!
2020-07-15 11:03:11 - INFO - chubaodb::master::server(43) - master listening on http://0.0.0.0:7070
2020-07-15 11:03:11 - INFO - actix_server::builder(262) - Starting 8 workers
2020-07-15 11:03:11 - INFO - actix_server::builder(276) - Starting "actix-web-service-0.0.0.0:7070" service on 0.0.0.0:7070
^C2020-07-15 11:03:12 - INFO - actix_server::builder(321) - SIGINT received, exiting
````


* 比如我们只启动 pserver 那么就是. `./chubaodb ps -c ../../config/config.toml` . ps会通过你配置文件中master的配置自动去master 注册为数据节点

````
load config by path: ../../config/config.toml
2020-07-15 11:03:45 - INFO - chubaodb::util::config(189) - log init ok 
2020-07-15 11:03:45 - INFO - chubaodb(149) - All ChubaoDB servers were started successfully!
2020-07-15 11:03:45 - INFO - chubaodb::util::http_client(21) - send get for url:http://127.0.0.1:7070/my_ip
2020-07-15 11:03:45 - INFO - surf::middleware::logger::native(119) - sending request
2020-07-15 11:03:45 - WARN - isahc::handler(209) - request completed with error [id=AtomicCell { value: 0 }]: ConnectFailed: failed to connect to the server
````



src/util/config.rs
// Copyright 2020 The Chubao Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
// implied. See the License for the specific language governing
// permissions and limitations under the License.
use crate::util::net::MyIp;
use git_version::git_version;
use log::{info, LevelFilter};
use log4rs::{
    append::{
        console::{ConsoleAppender, Target},
        rolling_file::{
            policy::compound::{
                roll::fixed_window::FixedWindowRoller, trigger::size::SizeTrigger, CompoundPolicy,
            },
            RollingFileAppender,
        },
    },
    config::{Appender, Config as LogConfig, Root},
    encode::pattern::PatternEncoder,
    filter::threshold::ThresholdFilter,
};
use serde_derive::Deserialize;
use std::fs::File;
use std::io::prelude::*;
use toml;

pub const VERSION: &str = clap::crate_version!();
pub const GIT_VERSION: &str = git_version!();

#[derive(Debug, Deserialize, Clone)]
pub struct Config {
    pub global: Global,
    pub router: Router,
    pub ps: PS,
    pub masters: Vec<Master>,
}

#[derive(Debug, Deserialize, Clone)]
pub struct Global {
    pub name: String,
    #[serde(default = "empty_str")]
    pub ip: String,
    pub log: String,
    pub log_level: String,
    #[serde(default = "default_log_limit_bytes")]
    pub log_limit_bytes: usize,
    #[serde(default = "default_log_file_count")]
    pub log_file_count: usize,
    pub shared_disk: bool,
}

fn default_log_limit_bytes() -> usize {
    128 * 1024 * 1024
}

fn default_log_file_count() -> usize {
    100
}

#[derive(Debug, Deserialize, Clone)]
pub struct Router {
    pub http_port: u16,
}

#[derive(Debug, Deserialize, Clone)]
pub struct PS {
    // id value not need set in config, It will be assigned by the master
    pub id: Option<u64>,
    pub zone: String,
    pub data: String,
    pub rpc_port: u16,
    pub flush_sleep_sec: Option<u64>,
    pub raft: RaftConf,
}

#[derive(Debug, Deserialize, Clone)]
pub struct RaftConf {
    pub heartbeat_port: u16,
    pub replicate_port: u16,
    // how size of num for memory
    pub log_max_num: usize,
    // how size of num for memory
    pub log_min_num: usize,
    // how size of num for memory
    pub log_file_size_mb: u64,
    //Three  without a heartbeat , follower to begin consecutive elections
    pub heartbeate_ms: u64,
}

#[derive(Debug, Deserialize, Clone)]
pub struct Master {
    pub ip: String,
    pub http_port: u16,
    #[serde(default = "false_bool")]
    pub is_self: bool,
    pub data: String,
}

impl Config {
    //init once for starup
    fn init(&mut self) {
        if self.global.ip == "" {
            let my = MyIp::instance().unwrap();
            for m in self.masters.iter_mut() {
                if my.is_my_ip(m.ip.as_str()) {
                    m.is_self = true;
                    break;
                }
            }
        } else {
            for m in self.masters.iter_mut() {
                if m.ip.as_str() == self.global.ip.as_str() {
                    m.is_self = true;
                    break;
                }
            }
        }

        // init log in
        let level = match self.global.log_level.to_uppercase().as_str() {
            "DEBUG" => {
                std::env::set_var("RUST_LOG", "actix_web=info"); //FIXME: not worked
                LevelFilter::Debug
            }
            "INFO" => LevelFilter::Info,
            "WARN" => LevelFilter::Warn,
            "TRACE" => LevelFilter::Trace,
            "ERROR" => LevelFilter::Error,
            "OFF" => LevelFilter::Off,
            _ => panic!("can not find log level:{}", self.global.log_level),
        };

        let stdout = ConsoleAppender::builder()
            .target(Target::Stdout)
            .encoder(Box::new(PatternEncoder::new(
                "{d(%Y-%m-%d %H:%M:%S)} - {l} - {t}\\({L}\\) - {m}{n}",
            )))
            .build();

        let chubaodb = RollingFileAppender::builder()
            .encoder(Box::new(PatternEncoder::new(
                "{d(%Y-%m-%d %H:%M:%S)} - {l} - {t}\\({L}\\) - {m}{n}",
            )))
            .build(
                std::path::Path::new(self.global.log.as_str()).join("chubaodb.log"),
                Box::new(CompoundPolicy::new(
                    Box::new(SizeTrigger::new(1024 * 1024 * 128)),
                    Box::new(
                        FixedWindowRoller::builder()
                            .build(
                                std::path::Path::new(self.global.log.as_str())
                                    .join("chubaodb.{}.log")
                                    .to_str()
                                    .unwrap(),
                                2000,
                            )
                            .unwrap(),
                    ),
                )),
            )
            .unwrap();

        let config = LogConfig::builder()
            .appender(Appender::builder().build("chubaodb", Box::new(chubaodb)))
            .appender(
                Appender::builder()
                    .filter(Box::new(ThresholdFilter::new(level)))
                    .build("stdout", Box::new(stdout)),
            )
            .build(
                Root::builder()
                    .appender("chubaodb")
                    .appender("stdout")
                    .build(level),
            )
            .unwrap();

        let _handle = log4rs::init_config(config).expect("init log config has err");

        info!("log init ok ");
    }

    pub fn self_master(&self) -> Option<Master> {
        for m in self.masters.iter() {
            if m.is_self {
                return Some(m.clone());
            }
        }
        None
    }

    pub fn master_addr(&self) -> String {
        return format!("{}:{}", self.masters[0].ip, self.masters[0].http_port);
    }
}

pub fn load_config(conf_path: &str, ip: Option<&str>) -> Config {
    let mut config = _load_config(conf_path, ip);
    config.init();
    config
}

fn _load_config(conf_path: &str, ip: Option<&str>) -> Config {
    if conf_path == "default" {
        return Config {
            global: Global {
                name: String::from("chubaodb"),
                ip: String::from("127.0.0.1"),
                log: String::from("log/"),
                log_level: String::from("info"),
                log_limit_bytes: default_log_limit_bytes(),
                log_file_count: default_log_file_count(),
                shared_disk: true,
            },
            ps: PS {
                id: None,
                zone: String::from("default"),
                data: String::from("data/ps"),
                rpc_port: 9090,
                flush_sleep_sec: Some(3),
                raft: RaftConf {
                    heartbeat_port: 12130,
                    replicate_port: 12131,
                    log_max_num: 20000,
                    log_min_num: 10000,
                    log_file_size_mb: 32,
                    heartbeate_ms: 500,
                },
            },
            router: Router { http_port: 8080 },
            masters: vec![Master {
                ip: String::from("127.0.0.1"),
                http_port: 7070,
                is_self: true,
                data: String::from("data/"),
            }],
        };
    }

    let mut file = match File::open(conf_path) {
        Ok(f) => f,
        Err(e) => panic!("no such file {} exception:{}", conf_path, e),
    };
    let mut str_val = String::new();

    match file.read_to_string(&mut str_val) {
        Ok(s) => s,
        Err(e) => panic!("Error Reading file: {}", e),
    };
    let mut config: Config = toml::from_str(&str_val).unwrap();

    if let Some(ip) = ip {
        config.global.ip = ip.to_string();
    }

    return config;
}

fn empty_str() -> String {
    "".to_string()
}

fn false_bool() -> bool {
    false
}
